core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1630,6 +1652,50 @@ private[spark] class DAGScheduler(
     }
   }
 
+  private[scheduler] def checkAndScheduleShuffleMergeFinalize(
+      shuffleStage: ShuffleMapStage): Unit = {
+    // Check if a finalize task has already been scheduled. This is to prevent scenarios
+    // where we don't schedule multiple shuffle merge finalization which can happen due
+    // stage retry or shufflePushMinRatio is already hit etc.
+    if (shuffleStage.shuffleDep.getFinalizeTask.isEmpty) {
+      // 1. Stage indeterminate and some map outputs are not available - finalize
+      // immediately without registering shuffle merge results.
+      // 2. Stage determinate and some map outputs are not available - decide to
+      // register merge results based on map outputs size available and
+      // shuffleMergeWaitMinSizeThreshold.
+      // 3. All shuffle outputs available - decide to register merge results based
+      // on map outputs size available and shuffleMergeWaitMinSizeThreshold.
+      val totalSize = {
+        lazy val computedTotalSize =
+          mapOutputTracker.getStatistics(shuffleStage.shuffleDep).
+            bytesByPartitionId.filter(_ > 0).sum
+        if (shuffleStage.isAvailable) {
+          computedTotalSize
+        } else {
+          if (shuffleStage.isIndeterminate) {
+            0L
+          } else {
+            computedTotalSize
+          }
+        }
+      }
+
+      if (totalSize < shuffleMergeWaitMinSizeThreshold) {
+        // Check if we can process map stage completion. If shuffle merge finalization
+        // is already triggered because push completion ratio was reached earlier,
+        // we cannot process map stage completion, but have to wait for the finalization
+        // to finish. This is because it's not straightforward to interrupt the
+        // finalization task and undo what it might have already done.
+        if (scheduleShuffleMergeFinalize(shuffleStage, delay = 0,
+          registerMergeResults = false)) {
```

Comment:
@Ngone51 Yes `scheduleShuffleMergeFinalize` would call `finalizeShuffleMerge` in a separate thread but the `handleShuffleMergeFinalized` should happen with in the `DAGSchedulerEventLoop` thread. Is the question that `handleShuffleMergeFinalized` happens outside of `DAGSchedulerEventLoop`? Or can you please clarify the question if it is different?

Reply:
Okay I see. Do you mean instead of calling `handleShuffleMergeFinalized` here we should just post `ShuffleMergeFinalized` and let the finalization happen inside `DAGSchedulerEventLoop` that is a good point. 

Just for context, in this case we are just marking the `ShuffleDependency` `shuffleMergeFinalized` to proceed with the next stage as the shuffle generated is < `shuffleMergeWaitMinSizeThreshold` so no need to wait for shuffle services to merge the shuffle merge output.

Reply:
No, I mean, when `scheduleShuffleMergeFinalize` returns `true`, we'll call `handleShuffleMergeFinalized` two times, which is not necessary. Right?

Reply:
Yes that is true but the advantage here is since the size of shuffle data generated is < `shuffleMergeWaitMinSizeThreshold` we want to immediately mark shuffle merge finalized so that the next stage would proceed immediately without waiting for all the shuffle mergers (external shuffle services) to finalize the shuffle merge. This is just an optimization to reduce the overall delay in proceeding with the next stage.

Reply:
Ok, I got your point.

I took another look at this part code. Seems like I misread one part. When `registerMergeResults=false`, it won't call `handleShuffleMergeFinalized` at the end (I thought it would before). So, it doesn't really call `handleShuffleMergeFinalized` in this case. 

But shall we refactor `finalizeShuffleMerge` to make it be consistent for `registerMergeResults=true|false`? e.g.,
```scala
...
val results = (0 until numMergers).map(_ => SettableFuture.create[Boolean]())
if (!registerMergeResults) {
  results.foreach(_.set(true))
} else {
  ...
}
try {
 Futures.allAsList(results: _*).get(shuffleMergeResultsTimeoutSec, TimeUnit.SECONDS)
} catch {
  case _: TimeoutException =>
    logInfo(s"Timed out on waiting for merge results from all " +
     s"$numMergers mergers for shuffle $shuffleId")
} finally {
  eventProcessLoop.post(ShuffleMergeFinalized(stage))
}
...

```

Reply:
Makes sense.